README.md
# HSegmentControl
[![CI Status](http://img.shields.io/travis/popodidi/HSegmentControl.svg?style=flat)](https://travis-ci.org/popodidi/HSegmentControl)
[![Version](https://img.shields.io/cocoapods/v/HSegmentControl.svg?style=flat)](http://cocoapods.org/pods/HSegmentControl)
[![License](https://img.shields.io/cocoapods/l/HSegmentControl.svg?style=flat)](http://cocoapods.org/pods/HSegmentControl)
[![Platform](https://img.shields.io/cocoapods/p/HSegmentControl.svg?style=flat)](http://cocoapods.org/pods/HSegmentControl)

A customized segment control sublassing UIControl where the view, titles of each segment and the indicator view can all be cutomized.

### Features

- [x] Customized configuration
	- `@IBInspectable`Title color
	- `@IBInspectable`Indicator image
	- `@IBInspectable`Number of displayed segments 
	- Title font
	- Indicator view
- [x] DataSource protocol
	- Number of segments
	- Segment background view (optional)
- [x] Documentation

### Demo

![](demo.gif)


## Example

To run the example project, clone the repo, and run `pod install` from the Example directory first.


## Installation

HSegmentControl is available through [CocoaPods](http://cocoapods.org). To install
it, simply add the following line to your Podfile:

```ruby
pod "HSegmentControl"
```

Swift version | version
---- | ----
2.3 | 2.0.x
3.0 | 2.1.0 or later

## Usage

### Add to View

You can either add `HSegmentControl` programmatically with code

``` swift
let segmentControl = HSegmentControl()
view.addSubview(segmentControl)
// Customized configuration
// .
// .
// .
```

or simply drag and drop an UIView in storyboard (Interface Builder) and set the class to `HSegmentControl`.

### Data source protocol
Make sure your controller implements the `HSegmentControlDataSource` protocol and is set to the `dataSource` of the `HSegmentControl`

```swift
class ViewController: UIViewController, HSegmentControlDataSource{
	// .
	// .
	// .
	override func viewDidLoad(){
		super.viewDidLoad()	
		segmentControl.dataSource = self
		//.
		//.
		//.
	}
	// .
	// .
	// .
	// MARK: - HSegmentControlDataSource protocol
    func numberOfSegments(segmentControl: HSegmentControl) -> Int {
        return 5
    }
    
    func segmentControl(segmentControl: HSegmentControl, titleOfIndex index: Int) -> String {
        return ["1","two", "threeeeeee", "four", "five"][index]
    }
    
    // Optional
    func segmentControl(segmentControl: HSegmentControl, segmentBackgroundViewOfIndex index: Int) -> UIView {
        let view = UIView()
        view.backgroundColor = UIColor(red: CGFloat(drand48()), green: CGFloat(drand48()), blue: CGFloat(drand48()), alpha: 1)
        return view
    }
}
```

### Configure HSegmentControl
```swift
override func viewDidLoad(){
	super.viewDidLoad()
	segmentControl.dataSource = self
	//.
	//.
	//.
	// Customized configuration
	segmentControl.numberOfDisplayedSegments = 3
	segmentControl.selectedTitleFont = UIFont.boldSystemFontOfSize(17)
	segmentControl.selectedTitleColor = UIColor.darkGrayColor()
	segmentControl.unselectedTitleFont = UIFont.systemFontOfSize(17)
	segmentControl.unselectedTitleColor = UIColor.blackColor()
	segmentControl.segmentIndicatorImage = UIImage(named: "ind_img")
	segmentControl.segmentIndicatorViewContentMode = UIViewContentMode.Bottom
	segmentControl.segmentIndicatorView.backgroundColor = UIColor.whiteColor()
}
```

### Value changed
As you did in any other `UIControl`, link the `valueChanged` functions to delegate action when value changed.

```swift
@IBAction func valueChanged(sender: AnyObject){
	print("value did change to \((sender as! HSegmentControl).selectedIndex)")
}
```

### Change value manually
Simply set the `selectedIndex` variable

```swift
segmentControl.selectedIndex = 2
```
Note that function `valueChanged(sender:)` will not be called when you set `selectedIndex` manually.

## Author

Hao


## License

HSegmentControl is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
# Change Log
### 2.1.2
- public `HSegmentControlDataSource` extension

### 2.1.1
- `HSegmentControlDatasSource` more swifty
- use `weak var` instead

### 2.1.0
- Swift 3

### 2.0.0
- Swift 2.3
- Documentation

### 1.0.7
- Fix total offset bug

### 1.0.6
- Update `README.md`

### 1.0.5
- Manual Deployment

### 1.0.4
- Travis CI auto deployment

### 1.0.3
- Try to fix Travis CI auto deployment

### 1.0.2
- Add Travis CI auto deployment
- Add Example/Pods to git
- Fix label color/font updating bug

### 1.0.1
- Add Travis CI

### 1.0
- First published version.


Example/Tests/Tests.swift
// https://github.com/Quick/Quick

import Quick
import Nimble
import HSegmentControl

class TableOfContentsSpec: QuickSpec {
//    override func spec() {
//        describe("these will fail") {
//
//            it("can do maths") {
//                expect(1) == 2
//            }
//
//            it("can read") {
//                expect("number") == "string"
//            }
//
//            it("will eventually fail") {
//                expect("time").toEventually( equal("done") )
//            }
//            
//            context("these will pass") {
//
//                it("can do maths") {
//                    expect(23) == 23
//                }
//
//                it("can read") {
//                    expect("🐮") == "🐮"
//                }
//
//                it("will eventually pass") {
//                    var time = "passing"
//
//                    dispatch_async(dispatch_get_main_queue()) {
//                        time = "done"
//                    }
//
//                    waitUntil { done in
//                        NSThread.sleepForTimeInterval(0.5)
//                        expect(time) == "done"
//
//                        done()
//                    }
//                }
//            }
//        }
//    }
}


Example/HSegmentControl/AppDelegate.swift
//
//  AppDelegate.swift
//  HSegmentControl
//
//  Created by Chang, Hao on 08/29/2016.
//  Copyright (c) 2016 Chang, Hao. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        return true
    }

    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



HSegmentControl/Classes/HSegmentControl.swift
//
//  HSegmentControl.swift
//  Pods
//
//  Created by Chang, Hao on 8/29/16.
//
//

import UIKit

@IBDesignable
open class HSegmentControl: UIControl {

    // MARK: - Data Source
    /**
     HSegmentControlDataSource
     */
    open weak var dataSource: HSegmentControlDataSource?{
        didSet{
            reloadData()
            if selectedIndex >= numberOfSegments {
                selectedIndex = numberOfSegments - 1
            }
        }
    }
    
    // MARK: - Calculated variables from data source
    /**
     Array of segment titles
     */
    open var segmentTitles: [String]{
        get{
            var titles = [String]()
            for index in 0 ..< numberOfSegments {
                titles.append(dataSource?.segmentControl(self, titleOfIndex: index) ?? "")
            }
            return titles
        }
    }
    fileprivate var numberOfSegments: Int{
        get{
            return dataSource?.numberOfSegments(self) ?? 0
        }
    }
    
    // MARK: - Control Status
    /**
     Selected Index of segment control
     */
    open var selectedIndex: Int = 0 {
        didSet{
            if selectedIndex < numberOfSegments{
                if selectedIndex != oldValue {
                    select(atIndex: selectedIndex, previousIndex: oldValue)
                }
            }else{
                selectedIndex = numberOfSegments - 1
            }
        }
    }
    
    // MARK: - Configuration
    /**
     The number of displayed segments with 3 as default value
     */
    @IBInspectable open var numberOfDisplayedSegments: Int = 3{
        didSet{
            reloadData()
        }
    }
    /**
     The indicator image of segment control with nil as default value
     */
    @IBInspectable open var segmentIndicatorImage: UIImage?{
        didSet{
            reloadData()
        }
    }
    /**
     The `ContentMode` of segment indicator image view
     */
    @IBInspectable open var segmentIndicatorViewContentMode: UIViewContentMode?{
        didSet{
            reloadData()
        }
    }
    /**
     The bakcgorundColor of segment indicator image view
     */
    @IBInspectable open var segmentIndicatorBackgroundColor: UIColor?{
        didSet{
            reloadData()
        }
    }
    /**
     The title color of unselected segment
     */
    @IBInspectable open var unselectedTitleColor: UIColor?{
        didSet{
            reloadData()
        }
    }
    /**
     The title color of selected segment
     */
    @IBInspectable open var selectedTitleColor: UIColor?{
        didSet{
            reloadData()
        }
    }
    /**
     The font of unselected segment with `UIFont.systemFontOfSize(UIFont.systemFontSize())` as default value
     */
    open var unselectedTitleFont: UIFont?{
        didSet{
            reloadData()
        }
    }
    /**
     The font of selected segment with `UIFont.systemFontOfSize(UIFont.systemFontSize())` as default value
     */
    open var selectedTitleFont: UIFont?{
        didSet{
            reloadData()
        }
    }
    fileprivate let defaultFont = UIFont.systemFont(ofSize: UIFont.systemFontSize)
    
    // MARK: - UI Elements
    open var segmentIndicatorView: UIView = UIView()
    
    fileprivate var baseView: UIView!
    fileprivate var segmentBaseViews: [UIView] = []
    fileprivate var segmentTitleLabels: [UILabel] = []
    
    // MARK: - Constraints for animation
    fileprivate var indicatorXConstraint = NSLayoutConstraint()
    fileprivate var selectedSegmentXContraint = NSLayoutConstraint()
    
    // MARK: - init
    override public init(frame: CGRect){
        super.init(frame: frame)
        configure()
        reloadData()
    }
    
    required public init?(coder: NSCoder) {
        super.init(coder: coder)
        configure()
        reloadData()
    }
    
    func configure(){
        clipsToBounds = true
    }
    
    // MARK: - reload data
    /**
     Reload the segment control
     */
    open func reloadData(){
        configureBaseView()
        configureSegmentBaseViews()
        configureIndicatorView()
        configureSegmentTitleLabels()
        layoutIfNeeded()
    }

    // MARK: - View Configuration
    func configureBaseView(){
        baseView = UIView(frame: CGRect(origin: CGPoint.zero, size: self.frame.size))
        baseView.translatesAutoresizingMaskIntoConstraints = false
        self.addSubview(baseView)
        
        let heightContraint = NSLayoutConstraint(item: baseView, attribute: .height, relatedBy: .equal, toItem: self, attribute: .height, multiplier: 1, constant: 0)
        let yContraint = NSLayoutConstraint(item: baseView, attribute: .centerY, relatedBy: .equal, toItem: self, attribute: .centerY, multiplier: 1, constant: 0)
        
        self.addConstraints([heightContraint, yContraint])
    }
    
    func configureSegmentBaseViews(){
        for segmentBaseView in segmentBaseViews{
            segmentBaseView.removeFromSuperview()
        }
        segmentBaseViews.removeAll(keepingCapacity: true)
        
        for index in 0 ..< numberOfSegments{
            let segmentBaseView = dataSource?.segmentControl(self, segmentBackgroundViewOfIndex: index) ?? UIView()
            segmentBaseView.translatesAutoresizingMaskIntoConstraints = false
            
            segmentBaseViews.append(segmentBaseView)
            baseView.addSubview(segmentBaseView)
            
            // width constraint
            self.addConstraint(NSLayoutConstraint(item: segmentBaseView, attribute: .width, relatedBy: .equal, toItem: self, attribute: .width, multiplier: 1/CGFloat(numberOfDisplayedSegments), constant: 0))
            // height constraint
            baseView.addConstraint(NSLayoutConstraint(item: segmentBaseView, attribute: .height, relatedBy: .equal, toItem: baseView, attribute: .height, multiplier: 1, constant: 0))
            // y constraint
            baseView.addConstraint(NSLayoutConstraint(item: segmentBaseView, attribute: .centerY, relatedBy: .equal, toItem: baseView, attribute: .centerY, multiplier: 1, constant: 0))
            
            // leading constraint
            if index == 0 {
                baseView.addConstraint(NSLayoutConstraint(item: segmentBaseView, attribute: .leading, relatedBy: .equal, toItem: baseView, attribute: .leading, multiplier: 1, constant: 0))
            }else{
                baseView.addConstraint(NSLayoutConstraint(item: segmentBaseView, attribute: .leading, relatedBy: .equal, toItem: segmentBaseViews[index-1], attribute: .trailing, multiplier: 1, constant: 0))
            }
            
            // trailing constraint
            if index == numberOfSegments - 1 {
                baseView.addConstraint(NSLayoutConstraint(item: segmentBaseView, attribute: .trailing, relatedBy: .equal, toItem: baseView, attribute: .trailing, multiplier: 1, constant: 0))
            }
            
            // total x constraint
            if index == selectedIndex {
                self.removeConstraint(selectedSegmentXContraint)
                if index == 0{
                    selectedSegmentXContraint = NSLayoutConstraint(item: segmentBaseView, attribute: .leading, relatedBy: .equal, toItem: self, attribute: .leading, multiplier: 1, constant: 0)
                }else if index == numberOfSegments - 1 {
                    selectedSegmentXContraint = NSLayoutConstraint(item: segmentBaseView, attribute: .trailing, relatedBy: .equal, toItem: self, attribute: .trailing, multiplier: 1, constant: 0)
                }else{
                    selectedSegmentXContraint = NSLayoutConstraint(item: segmentBaseView, attribute: .centerX, relatedBy: .equal, toItem: self, attribute: .centerX, multiplier: 1, constant: 0)
                }
                self.addConstraint(selectedSegmentXContraint)
            }
            
        }
    }
    
    func configureIndicatorView(){
        segmentIndicatorView.removeFromSuperview()
        segmentIndicatorView = UIImageView()
        segmentIndicatorView.translatesAutoresizingMaskIntoConstraints = false
        (segmentIndicatorView as? UIImageView)?.image = segmentIndicatorImage
        (segmentIndicatorView as? UIImageView)?.contentMode = segmentIndicatorViewContentMode ?? .bottom
        
        baseView.addSubview(segmentIndicatorView)
        
        // width constraint
        self.addConstraint(NSLayoutConstraint(item: segmentIndicatorView, attribute: .width, relatedBy: .equal, toItem: self, attribute: .width, multiplier: 1/CGFloat(numberOfDisplayedSegments), constant: 0))
        // height constraint
        baseView.addConstraint(NSLayoutConstraint(item: segmentIndicatorView, attribute: .height, relatedBy: .equal, toItem: baseView, attribute: .height, multiplier: 1, constant: 0))
        // y constraint
        baseView.addConstraint(NSLayoutConstraint(item: segmentIndicatorView, attribute: .centerY, relatedBy: .equal, toItem: baseView, attribute: .centerY, multiplier: 1, constant: 0))
        
        
        setIndicatorXConstraint()
        
    }
    
    func configureSegmentTitleLabels(){
        for segmentTitleLabel in segmentTitleLabels{
            segmentTitleLabel.removeFromSuperview()
        }
        segmentTitleLabels.removeAll(keepingCapacity: true)
        
        for index in 0 ..< numberOfSegments{
            let titleLabel = UILabel()
            titleLabel.text = segmentTitles[index]
            titleLabel.textAlignment = .center
            titleLabel.textColor = index == selectedIndex ? selectedTitleColor : unselectedTitleColor
            
            titleLabel.font = (index == selectedIndex ? selectedTitleFont : unselectedTitleFont) ?? defaultFont
            titleLabel.translatesAutoresizingMaskIntoConstraints = false
            baseView.addSubview(titleLabel)
            segmentTitleLabels.append(titleLabel)
            
            let widthContraint = NSLayoutConstraint(item: titleLabel, attribute: .width, relatedBy: .equal, toItem: segmentBaseViews[index], attribute: .width, multiplier: 0.8, constant: 0)
            let heightContraint = NSLayoutConstraint(item: titleLabel, attribute: .height, relatedBy: .equal, toItem: segmentBaseViews[index], attribute: .height, multiplier: 0.8, constant: 0)
            let xContraint = NSLayoutConstraint(item: titleLabel, attribute: .centerX, relatedBy: .equal, toItem: segmentBaseViews[index], attribute: .centerX, multiplier: 1, constant: 0)
            let yContraint = NSLayoutConstraint(item: titleLabel, attribute: .centerY, relatedBy: .equal, toItem: segmentBaseViews[index], attribute: .centerY, multiplier: 1, constant: 0)
            
            baseView.addConstraints([widthContraint, heightContraint, xContraint, yContraint])
        }
    }
    
    // MARK: - Subclassing UIView
    override open func layoutSubviews() {
        super.layoutSubviews()
    }
    
    override open func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        super.touchesBegan(touches, with: event)
        guard let location = touches.first?.location(in: baseView) else{
            return
        }
        guard let calculatedIndex = { () -> Int? in
            for (index, segmentBaseView) in segmentBaseViews.enumerated() {
                if segmentBaseView.frame.contains(location) {
                    return index
                }
            }
            return nil
            }()
            else{
                return
        }
        
        if selectedIndex != calculatedIndex{
            selectedIndex = calculatedIndex
            sendActions(for: .valueChanged)
        }
    }
    
    // MARK: - Animation
    func select(atIndex nextIndex: Int, previousIndex preIndex: Int){
        UIView.animate(withDuration: 0.1, animations: {
            if preIndex < self.numberOfSegments{
                for label in self.segmentTitleLabels{
                    label.textColor = self.unselectedTitleColor
                    label.font = self.unselectedTitleFont ?? self.defaultFont
                }
            }
        }, completion: { (completed) in
            UIView.animate(withDuration: 0.2, delay: 0, usingSpringWithDamping: 0.7, initialSpringVelocity: 0.6, options: .curveEaseOut, animations: {
                self.setTotalXConstraint()
                self.setIndicatorXConstraint()
                self.addConstraint(self.selectedSegmentXContraint)
                self.layoutIfNeeded()
            }){ (completed) in
                UIView.animate(withDuration: 0.1, animations: {
                    if nextIndex < self.numberOfSegments{
                        self.segmentTitleLabels[nextIndex].textColor = self.selectedTitleColor
                        self.segmentTitleLabels[nextIndex].font = self.selectedTitleFont ?? self.defaultFont
                    }
                })
            }
        }) 
    }
    
    // MARK: - Set x position constraints
    func setTotalXConstraint(){
        if numberOfSegments > numberOfDisplayedSegments
            && selectedIndex < numberOfSegments
            && segmentBaseViews.count > 0{
            // total x constraint
            self.removeConstraint(self.selectedSegmentXContraint)
            if CGFloat(selectedIndex) < CGFloat(self.numberOfDisplayedSegments)/2{
                self.selectedSegmentXContraint = NSLayoutConstraint(item: self.segmentBaseViews.first!, attribute: .leading, relatedBy: .equal, toItem: self, attribute: .leading, multiplier: 1, constant: 0)
            }else if CGFloat(selectedIndex) > CGFloat(self.numberOfSegments) - CGFloat(self.numberOfDisplayedSegments)/2 - 1 {
                self.selectedSegmentXContraint = NSLayoutConstraint(item: self.segmentBaseViews.last!, attribute: .trailing, relatedBy: .equal, toItem: self, attribute: .trailing, multiplier: 1, constant: 0)
            }else{
                self.selectedSegmentXContraint = NSLayoutConstraint(item: self.segmentBaseViews[selectedIndex], attribute: .centerX, relatedBy: .equal, toItem: self, attribute: .centerX, multiplier: 1, constant: 0)
            }
        }
    }
    
    func setIndicatorXConstraint(){
        if selectedIndex < self.numberOfSegments {
            self.baseView.removeConstraint(self.indicatorXConstraint)
            self.indicatorXConstraint = NSLayoutConstraint(item: self.segmentIndicatorView, attribute: .centerX, relatedBy: .equal, toItem: self.segmentBaseViews[selectedIndex], attribute: .centerX, multiplier: 1, constant: 0)
            self.baseView.addConstraint(self.indicatorXConstraint)
        }
    }
}


HSegmentControl/Classes/HSegmentControlDataSource.swift
//
//  HSegmentControlDataSource.swift
//  Pods
//
//  Created by Chang, Hao on 8/30/16.
//
//

import Foundation

public protocol HSegmentControlDataSource: class {
    /**
     Retrun the total number of segments
     */
    func numberOfSegments(_ segmentControl: HSegmentControl) -> Int
    /**
     Return the title of each segment
     */
    func segmentControl(_ segmentControl: HSegmentControl, titleOfIndex index: Int) -> String
    /**
     Return background view of each segment.
     The background view will be set with constraints to fit the size of each segment and will be set to `UIView()` if this method is not implemented.
     */
    func segmentControl(_ segmentControl: HSegmentControl, segmentBackgroundViewOfIndex index: Int) -> UIView
}

public extension HSegmentControlDataSource {
    func segmentControl(_ segmentControl: HSegmentControl, segmentBackgroundViewOfIndex index: Int) -> UIView {
        return UIView()
    }
}


Example/HSegmentControl/ViewController.swift
//
//  ViewController.swift
//  HSegmentControl
//
//  Created by Chang, Hao on 08/29/2016.
//  Copyright (c) 2016 Chang, Hao. All rights reserved.
//

import UIKit
import HSegmentControl

class ViewController: UIViewController, HSegmentControlDataSource {

    @IBOutlet weak var segmentControl: HSegmentControl!
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        segmentControl.dataSource = self
        segmentControl.numberOfDisplayedSegments = 4
        segmentControl.segmentIndicatorViewContentMode = UIViewContentMode.bottom
        segmentControl.selectedTitleFont = UIFont.boldSystemFont(ofSize: 17)
        segmentControl.selectedTitleColor = UIColor(red: 232/255, green: 76/255, blue: 86/255, alpha: 1)
        segmentControl.unselectedTitleFont = UIFont.systemFont(ofSize: 17)
        segmentControl.unselectedTitleColor = UIColor.darkGray
        segmentControl.segmentIndicatorImage = UIImage(named: "ind_img")
        segmentControl.segmentIndicatorView.backgroundColor = UIColor.white
    }

    // MARK: - HSegmentControlDataSource protocol
    func numberOfSegments(_ segmentControl: HSegmentControl) -> Int {
        return 10
    }
    
    func segmentControl(_ segmentControl: HSegmentControl, titleOfIndex index: Int) -> String {
        return ["1","two", "threeeeeee", "four", "five", "1","two", "threeeeeee", "four", "five"][index]
    }
    
    
    func segmentControl(_ segmentControl: HSegmentControl, segmentBackgroundViewOfIndex index: Int) -> UIView {
        let view = UIView()
        view.backgroundColor = UIColor(red: CGFloat(drand48()), green: CGFloat(drand48()), blue: CGFloat(drand48()), alpha: 1)
        return view
    }
    
    @IBAction func valueChanged(_ sender: AnyObject) {
        print("value did change to \((sender as! HSegmentControl).selectedIndex)")
    }

}

